lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm

```diff
@@ -0,0 +1,180 @@
+ Optional
+  instdata.run_name      => The run name or id.
+
+HELP
+}
+
+sub entity_types {
+    return (qw/ individual sample library instdata/);
+}
+
+sub _open_file_parser {
+    my $self = shift;
+
+    my $file = $self->file;
+    my ($dir, $basename, $ext) = File::Basename::fileparse($file, 'csv', 'tsv');
+    die $self->error_message("Cannot determine type for file: %s. It needs to end with .csv or .tsv.", $file) if not $ext;
+    my $parser = Text::CSV->new({
+            sep_char => ( $ext eq 'csv' ? ',' : "\t" ),
+            empty_is_undef => 1,
+        });
+    die $self->error_message('Failed to create Text::CSV parser!') if not $parser;
+
+    my $fh = Genome::Sys->open_file_for_reading($file);
+    my $headers = $parser->getline($fh);
+    die $self->error_message('File (%s) does not have any headers! Is it empty?', $file) if not $headers;
+    $parser->column_names($headers);
+
+    my $entity_attributes_ok = $self->_resolve_headers($headers);
+    return if not $entity_attributes_ok;
+
+    return sub{ return $parser->getline_hr($fh); };
+}
+
+sub _resolve_headers {
+    my ($self, $headers) = Params::Validate::validate_pos(@_, {type => HASHREF}, {type => ARRAYREF});
+
+    my @entity_attributes;
+    for my $header ( @$headers ) {
+        my ($type, $attribute) = split(/\./, $header, 2);
+        next if not defined $attribute; # silently skip columns w/o an entity type
+        die $self->error_message('Invalid entity type: %s', $type) if not List::MoreUtils::any { $type eq $_ } $self->entity_types;
+        push @entity_attributes, {
+            header => $header,
+            type => $type,
+            attribute => $attribute,
+        };
+    }
+    $self->_entity_attributes(\@entity_attributes);
+
+    return 1;
+}
+
+sub _resolve_entity_params_for_values {
+    my ($self, $line_ref) = Params::Validate::validate_pos(@_, {type => HASHREF}, {type => HASHREF});
+
+    my %entity_params = map { $_ => {} } $self->entity_types;
+    for my $entity_attribute ( @{$self->_entity_attributes} ) {
+        my $value = $line_ref->{ $entity_attribute->{header} };
+        next if not defined $value;
+        $entity_params{ $entity_attribute->{type} }->{ $entity_attribute->{attribute} } = $value;
+    }
+
+    return \%entity_params;
+}
+
+sub _resolve_names_for_entities {
+    my ($self, $entity_params) = Params::Validate::validate_pos(@_, {type => HASHREF}, {type => HASHREF});
+
+    my $nomenclature = delete $entity_params->{sample}->{nomenclature};
+    my $individual_name_part = delete $entity_params->{individual}->{name_part};
+    my $sample_name_part = delete $entity_params->{sample}->{name_part};
+
+    # sample
+    my $sample_name = $entity_params->{sample}->{name};
+    if ( $sample_name ) {
+        my @tokens = split(/\-/, $sample_name);
+        if ( @tokens <  3 ) {
+            die $self->error_message('Invalid sample name: %s. It must have at least 3 parts separated by dashes.', $sample_name);
```

Comment:
If this is an existing sample in the db, is it possible that the sample name would not conform to the spec?